### ISSUE TITLE
Regularization images not being used as regularizers

### ISSUE THREAD

**Issue by rockerBOO** [kohya-ss/sd-scripts#449]
Thank you for these scripts. They have been helpful to me and have learned a lot through the process. Thank you. 

I'm finding a lot of training with regularization images not producing very good results. Having poor success when suggesting it to others as well and makes it hard to recommend to others. 

I spent some time looking through the code to figure out how regularization images would be applied in the code. According to the dreambooth method, there is some comparison between the regularization image and the training image. I looked in the code and did not see this comparison. 

The `images` is from looking through the bucket image index by the `bucket_batch_size` https://github.com/kohya-ss/sd-scripts/blob/main/library/train_util.py#L913. But this also doesn't indicate it's using the regularization images. 

```python
for _ in bucket[image_index : image_index + bucket_batch_size]:
```

In my trials, including making regularization images of a bucket size, that it never added the regularization images to be added, nor had an indication of which images were regularization.

So I would be assuming that regularization images are not being used. Currently, when printing out the "captions" for the current trained images it's saying the class name for the regularization image and seems to be just using regularization images as part of the training. Using regularization images can double your training steps as well. That's because it's training on the regularization images from my experience.

Comparing with Hugging Face implementation of Dreambooth they have some sections that may be helpful.

From: https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth.py#L966-L980. Describes comparing the loss from the reg (prior) images and training images.  Compare to https://github.com/kohya-ss/sd-scripts/blob/v0.6.2/train_network.py#L633

```python
                if args.with_prior_preservation:
                    # Chunk the noise and model_pred into two parts and compute the loss on each part separately.
                    model_pred, model_pred_prior = torch.chunk(model_pred, 2, dim=0)
                    target, target_prior = torch.chunk(target, 2, dim=0)

                    # Compute instance loss
                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")

                    # Compute prior loss
                    prior_loss = F.mse_loss(model_pred_prior.float(), target_prior.float(), reduction="mean")

                    # Add the prior loss to the instance loss.
                    loss = loss + args.prior_loss_weight * prior_loss
                else:
                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")
```

And it seems to use this code with the training image, and the regularization images and then stack together. https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth.py#L453-L558 Compare to https://github.com/kohya-ss/sd-scripts/blob/v0.6.2/library/train_util.py#L992-L993

```python
    input_ids = [example["instance_prompt_ids"] for example in examples]
    pixel_values = [example["instance_images"] for example in examples]

    # Concat class and instance examples for prior preservation.
    # We do this to avoid doing two forward passes.
    if with_prior_preservation:
        input_ids += [example["class_prompt_ids"] for example in examples]
        pixel_values += [example["class_images"] for example in examples]

    pixel_values = torch.stack(pixel_values)
    pixel_values = pixel_values.to(memory_format=torch.contiguous_format).float()
```

And in this repo there's something like that in https://github.com/kohya-ss/sd-scripts/blob/v0.6.2/library/train_util.py#L991-L996, but that doesn't include a regularization image to compare with. 

```python
            images = torch.stack(images)
            images = images.to(memory_format=torch.contiguous_format).float()
```

Am I off base in this assessment? I want to be able to make a PR to resolve this, but would take some time to update the code. 

Thank you

**Comment by rockerBOO** [comment]
Thank you for your detailed response. Helping me learn once again. I see what you have done here, and I understand it now. 

I was not understanding how this regularization worked (still learning) and now it makes more sense. Re-read the Dreambooth paper to get better context again. Now I understand how a simple thing this regularization ends up being. 

I also understand how the order and usage of the regularization might produce different results due to showing different numbers of regularization and training images, especially in early steps. There were examples where the regularization images were showing prevalent in the early steps of the training. This made me question if it was working properly, but now I feel I understand this enough to analyze it appropriately. 

So the flow is we load up all the dataset/subsets into the data loader, the data loader shuffles them all together. Then we train by the batch size amount of images (including any number of training or regularization images). If it's a regularization image, we multiply it by the `prior_loss_weight` argument the user can set, defaulting to the suggested 1.0 which uses the full loss value from the reg images. 

I wonder if an alternating data loader (one training, one regularization image) and no shuffle would be a compromise with a batch size divisible by 2. This way it flows more evenly through the training, and can learn together on the same step before the model backwards pass. I am not sure if this is worthwhile to try.

Thank you again for your time, and your great explanation. I appreciate it.
